File: go/internal/encoding/base62_codec.go
package encoding

// Base62Codec converts unsigned integers into fixed-width Base62 strings using a
// right-to-left modulo loop that yields natural front zero-padding.
type Base62Codec struct {
	base62Alphabet *Alphabet
}

// NewBase62Codec builds a Base62Codec bound to the given Base62 alphabet.
func NewBase62Codec(base62Alphabet *Alphabet) *Base62Codec {
	return &Base62Codec{base62Alphabet: base62Alphabet}
}

// Encode renders the given value as a Base62 string of exactly the requested width,
// filling positions from the least significant digit backward.
func (base62Codec *Base62Codec) Encode(value uint32, width int) string {
	var radix uint32 = uint32(base62Codec.base62Alphabet.Size())
	var characters []byte = make([]byte, width)
	var remaining uint32 = value

	var position int

	for position = width - 1; position >= 0; position-- {
		var remainder uint32 = remaining % radix
		characters[position] = base62Codec.base62Alphabet.CharacterAt(int(remainder))
		remaining = remaining / radix
	}

	return string(characters)
}

File: ts/src/encoding/Base62Codec.ts
import { Alphabet } from "./Alphabet";


/**
 * Base62Codec converts unsigned integers into fixed-width Base62 strings using a right-to-left
 * modulo loop that yields natural front zero-padding.
 */
export class Base62Codec {
    private readonly _base62Alphabet: Alphabet;

    /**
     * Constructs a Base62Codec bound to the given Base62 alphabet.
     */
    public constructor(base62Alphabet: Alphabet) {
        this._base62Alphabet = base62Alphabet;
    }

    /**
     * Renders the given 32-bit value as a Base62 string of exactly the requested width, filling
     * positions from the least significant digit backward.
     */
    public encode(value: number, width: number): string {
        const radix: number = this._base62Alphabet.size();
        const characters: string[] = new Array<string>(width);

        let remaining: number = value >>> 0;

        for (let position: number = width - 1; position >= 0; position--) {
            const remainder: number = remaining % radix;
            characters[position] = this._base62Alphabet.characterAt(remainder);
            remaining = Math.floor(remaining / radix);
        }

        return characters.join("");
    }
}
